### DISCUSSION TITLE
Please explain why I got infinite loop.

### DISCUSSION THREAD

**Discussion by 39george** [aembke/fred.rs#190] [Q&A] [answered]
Hello, and thanks for this amazing crate!
I can't understand why I got an infinite loop here:
```rust
    while let Some(mut page) = redis_client
        .scan("*", Some(10), None)
        .try_next()
        .await
        .unwrap()
    {
        if let Some(keys) = page.take_results() {
            dbg!(keys);
        }
        page.next().unwrap();
    }
```
But if I extract `redis_client.scan(..)` into the variable, the infinite loop is gone, and I got all my keys from redis:
```rust
    let mut scan = redis_client.scan("*", Some(10), None);
    while let Some(mut page) = scan.try_next().await.unwrap() {
        if let Some(keys) = page.take_results() {
            dbg!(keys);
        }
        page.next().unwrap();
    }
```

**Comment by aembke** [comment, marked as answer]
Hi @ghashy, 

SCAN works by repeatedly sending a `SCAN <cursor> ...` command to the server with an updated cursor after each page. 

In the first example you're sending a new `SCAN 0 ...` call with a fresh starting cursor on each iteration in the while loop, and the second one is properly paging through the results. Or in other words - in the first example you're calling both `scan` and `try_next` on each loop iteration, but in the second you're only calling `try_next` on each iteration.

**Reply by 39george** [reply]
Thanks!
